strings/base_reference_produce.h

```diff
@@ -261,6 +261,95 @@ WINRT_EXPORT namespace winrt::Windows::Foundation
     }
 }
 
+namespace winrt::impl
+{
+    template <typename T, typename From>
+    T unbox_value_type(From&& value)
+    {
+        static_assert(!is_com_interface_v<T>);
+
+        if (!value)
+        {
+            throw hresult_no_interface();
+        }
+        if constexpr (std::is_enum_v<T>)
+        {
+            if (auto temp = value.try_as<Windows::Foundation::IReference<T>>())
+            {
+                return temp.Value();
+            }
+            else
+            {
+                return static_cast<T>(value.as<Windows::Foundation::IReference<std::underlying_type_t<T>>>().Value());
+            }
+        }
+#ifdef WINRT_IMPL_IUNKNOWN_DEFINED
+        else if constexpr (std::is_same_v<T, GUID>)
+        {
+            return value.as<Windows::Foundation::IReference<guid>>().Value();
+        }
+#endif
+        else
+        {
+            return value.as<Windows::Foundation::IReference<T>>().Value();
+        }
+    }
+
+    template <typename T, typename Ret = T, typename From, typename U>
+    Ret unbox_value_type_or(From&& value, U&& default_value)
```

Comment:
What do you think about a defaulted version for `default_value`, so someone can say `unbox_value_type_or_default<int32_t>` directly, instead of passing in the default value?

Reply:
All WinRT types have an "all zero bytes" representation for their default values, but I'm not sure about adding another function for this though. How common is it to want that default when that default isn't normally a usable value? And you can obviously always request that as the default.